### PR TITLE
Add retry(upTo:interval:on:)

### PR DESCRIPTION
**Pull request by tjnet**
#### Checklist

- [x] Updated CHANGELOG.md.

#### Summary

I implemented #224 and write test code

I'm new to this. Is this correct? 
Please review in your free time.

### CODE REVIEW

File: Tests/ReactiveSwiftTests/SignalProducerSpec.swift [hunk 1]
```diff
@@ -1962,6 +1962,59 @@ class SignalProducerSpec: QuickSpec {
...
+						} else {
+							scheduler.schedule { observer.sendCompleted() }
+						}
+						count += 1
+					}
+
+					var values: [Int] = []
+					var completed = false
+					
+					original.retry(upTo: Int.max, interval: 1, on: scheduler)
```
Comment (mdiep) [inline Tests/ReactiveSwiftTests/SignalProducerSpec.swift]: Please add a test that tests hitting the `uptown` limit.

File: Sources/SignalProducer.swift [hunk 2]
```diff
@@ -1817,6 +1817,39 @@ extension SignalProducer {
...
+			var producer = SignalProducer<Value, Error>(error: error)
+			if retries > 0 {
+				producer = SignalProducer.empty
+					.delay(interval, on: scheduler)
+					.concat(producer)
+			}
+			
+			retries -= 1
+			return producer
+		}.retry(upTo: count)
```
Comment (mdiep) [inline Sources/SignalProducer.swift]: I don't believe this is correct. I think the `.retry` should be inside the `flatMapError`.

If you add a test like I suggested below, this should become clear.

File: Tests/ReactiveSwiftTests/SignalProducerSpec.swift [hunk 3]
```diff
@@ -1962,6 +1962,59 @@ class SignalProducerSpec: QuickSpec {
 				let result = producer.single()
 				expect(result?.value) == 1
 			}
+			
+			context("with interval") {
+				
+				it("should send values at the given interval until hitting the limitation") {
```
Comment (mdiep) [inline Tests/ReactiveSwiftTests/SignalProducerSpec.swift]: This isn't a very accurate description of the test. It's testing the retrying, not testing the limitation.

File: Tests/ReactiveSwiftTests/SignalProducerSpec.swift [hunk 4]
```diff
@@ -1962,6 +1962,59 @@ class SignalProducerSpec: QuickSpec {
...
+							scheduler.schedule { observer.sendCompleted() }
+						}
+						count += 1
+					}
+
+					var values: [Int] = []
+					var completed = false
+					
+					original.retry(upTo: Int.max, interval: 1, on: scheduler)
+					.start { event in
```
Comment (mdiep) [inline Tests/ReactiveSwiftTests/SignalProducerSpec.swift]: This indentation is off. The ReactiveSwift style is to have methods like this be indented one stop.

File: Sources/SignalProducer.swift (hunk 2 above)
Comment (mdiep) [inline Sources/SignalProducer.swift]: This indentation doesn't match the ReactiveSwift style. It should be indented like so:

```swift
return flatMapError { error in
		…
	}
	.retry(upTo: count)
```

**Comment by mdiep** [comment]
Thanks for opening a PR!

File: Tests/ReactiveSwiftTests/SignalProducerSpec.swift (hunk 1 above)
Comment (tjnet) [inline Tests/ReactiveSwiftTests/SignalProducerSpec.swift]: Thank you for your review so much! I 'll add test!

File: Sources/SignalProducer.swift (hunk 2 above)
Comment (tjnet) [inline Sources/SignalProducer.swift]: Thank you for your review ! I 'll fix . 😄

File: Tests/ReactiveSwiftTests/SignalProducerSpec.swift (hunk 1 above)
Comment (tjnet) [inline Tests/ReactiveSwiftTests/SignalProducerSpec.swift]: I added a test that tests hitting the ``upto`` limit.
please see 
https://github.com/ReactiveCocoa/ReactiveSwift/pull/496/files#diff-9cd786f9e52e786771d96cf8c42d3befR2018

File: Sources/SignalProducer.swift (hunk 2 above)
Comment (tjnet) [inline Sources/SignalProducer.swift]: I fixed indentation.

File: Tests/ReactiveSwiftTests/SignalProducerSpec.swift (hunk 3 above)
Comment (tjnet) [inline Tests/ReactiveSwiftTests/SignalProducerSpec.swift]: I see. I updated description of the test. 😄

File: Tests/ReactiveSwiftTests/SignalProducerSpec.swift (hunk 4 above)
Comment (tjnet) [inline Tests/ReactiveSwiftTests/SignalProducerSpec.swift]: I fixed indentation 😄

**Comment by tjnet** [comment]
I updated pull request. Any thoughts? 
Please review in your free time.

File: Sources/SignalProducer.swift [hunk 5]
```diff
@@ -1817,6 +1817,40 @@ extension SignalProducer {
...
+			if retries > 0 {
+				producer = SignalProducer.empty
+					.delay(interval, on: scheduler)
+					.concat(producer)
+			}
+			
+			retries -= 1
+			return producer
+		}
+		.retry(upTo: count)
```
Comment (ikesyo) [inline Sources/SignalProducer.swift]: https://github.com/ReactiveCocoa/ReactiveSwift/pull/496#discussion_r129294400

This is not fully addressed yet.

```swift
return flatMapError { error in
		…
	}
	.retry(upTo: count)
```

Comment (tjnet) [inline Sources/SignalProducer.swift]: Oh, sorry.. I 'll fix soon!

File: Sources/SignalProducer.swift [hunk 6]
```diff
@@ -1817,6 +1817,40 @@ extension SignalProducer {
...
+				if retries > 0 {
+					producer = SignalProducer.empty
+						.delay(interval, on: scheduler)
+						.concat(producer)
+				}
+			
+				retries -= 1
+				return producer
+		}
+		.retry(upTo: count)
```
Comment (ikesyo) [inline Sources/SignalProducer.swift]: Please indent the two lines above as well.

Comment (tjnet) [inline Sources/SignalProducer.swift]: I sent pull request again. Is this correct?
7a5b7b9

File: Tests/ReactiveSwiftTests/SignalProducerSpec.swift [hunk 7]
```diff
@@ -1962,6 +1962,100 @@ class SignalProducerSpec: QuickSpec {
...
+					expect(count) == 2
+					expect(values) == [1, 2]
+					
+					scheduler.advance(by: .seconds(1))
+					expect(count) == 3
+					expect(values) == [1, 2, 3]
+					
+					scheduler.advance(by: .seconds(1))
+					expect(count) == 3
+					expect(values) == [1, 2, 3]
```
Comment (mdiep) [inline Tests/ReactiveSwiftTests/SignalProducerSpec.swift]: Can we verify that the `error` was sent as well?

Comment (tjnet) [inline Tests/ReactiveSwiftTests/SignalProducerSpec.swift]: I got it.
I 'll try to add testing of error and update CHANGELOG.
thanks!

Comment (tjnet) [inline Tests/ReactiveSwiftTests/SignalProducerSpec.swift]: I added

Comment (tjnet) [inline Tests/ReactiveSwiftTests/SignalProducerSpec.swift]: >Can we verify that the error was sent as well?

I added test. How about this commit https://github.com/ReactiveCocoa/ReactiveSwift/pull/496/commits/73d2d4d50018252e37b4d6aff287f75af9ee0572 ?

**Comment by mdiep** [comment]
Thanks for the PR! Sorry I've been slow reviewing. It's been a busy few weeks.

The Swift 4 linux build failed, but I think that's because it needs to be updated.

**Comment by tjnet** [comment]
Sorry for my late reply.. I deeply appreciate your review and feedback! 😄

**Pull request merged**